python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -60,74 +65,129 @@ cdef class Scalar:
         dtype : dtype
             A NumPy dtype.
         """
-
         value = cudf.utils.dtypes.to_cudf_compatible_scalar(value, dtype=dtype)
-
-        valid = value is not None
+        valid = not is_null_host_scalar(value)
 
         if dtype is None:
-            if value is None:
+            if not valid:
                 raise TypeError(
                     "dtype required when constructing a null scalar"
                 )
+            if isinstance(value, (np.datetime64, np.timedelta64)):
+                if np.isnat(value) and np.datetime_data(
+                    value.dtype
+                )[0] == 'generic':
+                    raise TypeError("Need a dtype to build a NaT Scalar")
+                else:
+                    dtype = value.dtype
             else:
                 dtype = value.dtype
-
         dtype = np.dtype(dtype)
 
+        # Caching Mechanism
+        self._initialize_cache(value, dtype)
+
+    def _initialize_cache(self, value, dtype):
+        if is_null_host_scalar(value):
+            self._host_value = cudf.NA
+        else:
+            self._host_value = value
+        # need to save the host dtype, because this determines
+        # the codepath for setting the device value later
+        self._host_dtype = dtype
+
+    def _host_value_current(self):
+        return self._host_value is not None
+
+    def _device_value_current(self):
+        return self.uptr._device_uptr != NULL
```

Comment:
Suggest renaming these to make it clear that it's checking if the value is current as opposed to returning the current value.
```suggestion
    def _is_host_value_current(self):
        return self._host_value is not None

    def _is_device_value_current(self):
        return self.uptr._device_uptr != NULL
```